### PR TITLE
nixos module for accepting licenses

### PR DESCRIPTION
**Pull request by ts468**
A general module where nixpkgs related licenses can be accepted.
The module also helps to documents the available options.

For example, the "allowUnfree" nixpkgs option can now be accepted
by the following configuration in the "nixos/configuration.nix".
  nixpkgs = {
    licenses = {
      allowUnfree = true;
    };
  };

The module also fixes the issue of passing "allowUnfree=true" to
the i686 package tree within x86_64 systems.

NOTE: The module only affects the evaluation of the system wide
      configuration. User initiated "nix-env" commands are not in-
      fluenced. A future "nixuser" program will take care of that.

As an example, the installation of oraclejdk is handled by the new
module.
The license of the oraclejdk{7,7psu,8} packages have been changed
to the "Oracle Binary Code License Agreement for the Java SE Platform
Products and JavaFX". The packages can now only be installed once the
license is accepted, e.g., within the "nixos/modules/misc/licenses.nix"
module.
To accept the license for nixos system wide evaluations, the following can
be added to the nixos/configuration.nix file:
  nixpkgs = {
    licenses = {
      oracleBinaryCodeLicenseForJava = "accept";
    };
  };

### CODE REVIEW

**Comment by edolstra** [comment]
Hm, I don't know. I don't really want us to start reproducing Nixpkgs config options in NixOS. I mean, you can already set
```
  nixpkgs.config.allowUnfree = true;
```

**Comment by ts468** [comment]
Please let me know what you think!

The need to better document license related options like, e.g., "allowUnfree" has been mentioned in #4493.
Also, by requiring to accept licenses explicitly, the concerns for making oraclejdk available through an automated download, as raised in #6557, should be addressed.

File: lib/licenses.nix [hunk 1]
```diff
@@ -347,6 +347,11 @@ lib.mapAttrs (n: v: v // { shortName = n; }) rec {
     fullName = "OpenSSL License";
   };
 
+  oraclejavabcl = {
+    fullName = "Oracle Binary Code License Agreement for the Java SE Platform Products and JavaFX";
+    url = "http://www.oracle.com/technetwork/java/javase/terms/license/index.html";
```
Comment (edolstra) [inline lib/licenses.nix]: Unfree licenses must have `free = false` to prevent them from being accepted by default.

**Comment by jcumming** [comment]
A naive, untested suggestion:

`myoraclejdk = pkgs.oraclejdk.override { license = accepted; };`

... or some mechanism *outside* of nixpkgs to override the `src` and the `license` attrs. That way nixpkgs has correct license information, and you can use a local repository of tarballs or whatever for the closed source, proprietary software.

**Comment by ts468** [comment]
@edolstra Hm, so what do you suggest, how should I expose, e.g., the oraclejavabcl to the user? Should I create an option in nixpkgs.config instead? Would that be better than going down the black-listing approach?
I understand that there is little benefit in what one can do from simply reproducing all nixpkgs.config options in NixOS, but sometimes I find it still very useful to have a proper NixOS option. For example, I very often consult http://nixos.org/nixos/options.html to check for options, and I think commonly used settings like, e.g., license related options should show up there as well. It's a great strength of NixOS to have such a powerful module system where, e.g., documented options can be centralized into one place, and with this PR I just try to extend the module system in that spirit---and of course to also make oraclejdk be downloaded automatically ;)

@jcumming Thank you very much for your thoughts, but ideally I want to get a solution for oraclejdk that works within nixpkgs. But maybe an integrated override option is the way to go!

**Comment by joachifm** [comment]
Is this still being worked on?

**Comment by ts468** [comment]
@joachifm I'm still interested in a solution, but it doesn't have a high priority for me at the moment.

**Comment by grahamc** [comment]
Closing for being very stale.

**Pull request closed without merging**